[READY] for Z debug. Industrial lift (not shuttle edition)

Dennok opened the pull request:
## About The Pull Request

<details>
  <summary>(spoiler) I try make something like this Hydraulic Industrial Lift</summary>

![Hydraulic Industrial Lift](https://5.imimg.com/data5/QH/TO/MY-2/945-500x500.jpg)

</details>

<details>
  <summary>(spoiler) Or scissor lift</summary>

![A mobile scissor lift, extended to near its highest position](https://upload.wikimedia.org/wikipedia/commons/thumb/6/62/Hebebuehne_Scissorlift.jpg/305px-Hebebuehne_Scissorlift.jpg)

</details>

Catwalk + Ladder + rolling_table = lift

Seems like it work now.

Ideas are welcome.

TODO:
Make something like lift control panel, console, controller. Now it controlled from platform.

## Why It's Good For The Game

Lifts cool.

## Changelog
:cl:
add: Lift platform that works like moving catwalk
add: Lift now on MultiZ debug station
/:cl:

Dennok committed:
Lattice lift

Dennok committed:
Update lattice.dm

Dennok committed:
up

Dennok commented on code/game/objects/structures/lattice.dm:
```diff
@@ -121,3 +121,181 @@
+	RegisterSignal(src, COMSIG_MOVABLE_UNCROSS, .proc/RemoveItemFromLift)
+
+/obj/structure/lattice/lift/proc/RemoveItemFromLift(datum/source, atom/movable/AM)
+	to_chat(world, "[src] RemoveItemFromLift([AM])")
```
This message shown twice when AM moves out of platform. Don't know why.

EdgeLordExe replied:
use debugger in vsc it should show you when it is called.

TiviPlus replied:
Well since it shows twice it's probably called twice

Have you tried changing to UNCROSSED and CROSSED?

Dennok replied:
Thanks, change to UNCROSSED can help.

Dennok committed:
Thange UNCROSS to UNCROSSED

Tharcoonvagh commented:
Can you build turrets on catwalks?
If so then this is going to break some stuff, oh boy

Dennok commented:
> Can you build turrets on catwalks?
> If so then this is going to break some stuff, oh boy

Why?

Dennok committed:
up

Dennok committed:
bad map up

Tharcoonvagh commented:
Its basically a tank, more so if you get an AI turret

Dennok committed:
mm map

Dennok committed:
up

Dennok commented:
Now lift dont catch constructed things.
Work on this.

OverlyGeneric commented:
Would emagging the lift allow you to crush people to death with the roof?

Dennok committed:
catch created atoms

Dennok committed:
Update industrial_lift.dm

Dennok committed:
Platform horizontal move

Dennok committed:
Update industrial_lift.dm

Dennok committed:
remove second lattice

Dennok committed:
up

Dennok committed:
add Destroy

Dennok committed:
newline

Dennok committed:
fix map

JenqaDev commented:
> Would emagging the lift allow you to crush people to death with the roof?

what roof?

Dennok commented:
Say something about this please.
@Rohesie @ShizCalev

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+///Collect and command
+/datum/lift_master
+	var/list/lift_platforms = list()
```
```suggestion
	var/list/lift_platforms
```
Turn this into a lazy list

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	var/list/lift_platforms = list()
+
+/datum/lift_master/Destroy()
+	for(var/obj/structure/industrial_lift/lift_platform in lift_platforms)
+		lift_platform.LMaster = null
```
```suggestion
	for(var/l in lift_platforms)
		var/obj/structure/industrial_lift/lift_platform = l
		lift_platform.LMaster = null
```
No need to type check

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+/datum/lift_master/Destroy()
+	for(var/obj/structure/industrial_lift/lift_platform in lift_platforms)
+		lift_platform.LMaster = null
+	lift_platforms.Cut()
```
```suggestion
	lift_platforms = null
```
lazylist

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	for(var/obj/structure/industrial_lift/lift_platform in lift_platforms)
+		lift_platform.LMaster = null
+	lift_platforms.Cut()
+	. = ..()
```
```suggestion
	return ..()
```
return at the end of the proc

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+
+///Collect all bordered platforms
+/datum/lift_master/proc/Rebuild_lift_plaform(obj/structure/industrial_lift/base_lift_platform)
+	lift_platforms |= base_lift_platform
```
```suggestion
	LAZYOR(lift_platforms, base_lift_platform)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+/datum/lift_master/proc/Rebuild_lift_plaform(obj/structure/industrial_lift/base_lift_platform)
+	lift_platforms |= base_lift_platform
+	var/list/possible_expansions = list(base_lift_platform)
+	while(possible_expansions.len)
```
```suggestion
	while(LAZYLEN(possible_expansions))
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	lift_platforms |= base_lift_platform
+	var/list/possible_expansions = list(base_lift_platform)
+	while(possible_expansions.len)
+		for(var/obj/structure/industrial_lift/borderline in possible_expansions)
+			var/list/result = borderline.lift_platform_expansion(src)
```
```suggestion
		for(var/b in possible_expansions)
			var/obj/structure/industrial_lift/borderline = b
			var/list/result = borderline.lift_platform_expansion(src)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	while(possible_expansions.len)
+		for(var/obj/structure/industrial_lift/borderline in possible_expansions)
+			var/list/result = borderline.lift_platform_expansion(src)
+			if(result && result.len)
```
```suggestion
			if(length(result))
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+		for(var/obj/structure/industrial_lift/borderline in possible_expansions)
+			var/list/result = borderline.lift_platform_expansion(src)
+			if(result && result.len)
+				for(var/obj/structure/industrial_lift/lift_platform in result)
+					if(!lift_platforms.Find(lift_platform))
+						lift_platform.LMaster = src
+						lift_platforms |= lift_platform
+						possible_expansions |= lift_platform
```
```suggestion
				for(var/p in result)
					if(lift_platforms.Find(p))
						continue
				var/obj/structure/industrial_lift/lift_platform = p
				lift_platform.LMaster = src
				lift_platforms |= lift_platform
				possible_expansions |= lift_platform
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+
+///Move all platforms together
+/datum/lift_master/proc/MoveLift(going, mob/user)
+	for(var/obj/structure/industrial_lift/lift_platform in lift_platforms)
+		lift_platform.travel(going)
```
```suggestion
	for(var/p in lift_platforms)
		var/obj/structure/industrial_lift/lift_platform = p
		lift_platform.travel(going)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	var/min_x = world.maxx
+	var/min_y = world.maxy
+	
+	for(var/obj/structure/industrial_lift/lift_platform in lift_platforms)
+		max_x = max(max_x, lift_platform.x)
```
```suggestion
	for(var/p in lift_platforms)
		var/obj/structure/industrial_lift/lift_platform = p
		max_x = max(max_x, lift_platform.x)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	//This must be safe way to border tile to tile move of bordered platforms, that excludes platform overlapping.
+	if( going & WEST )
+		//Go along the X axis from min to max, from left to right
+		for(var/x = min_x; x <= max_x; x++)
```
```suggestion
		for(var/x in min_x to max_x)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+					lift_platform.travel(going)	
+	else
+		//Go along the X axis from max to min, from right to left
+		for(var/x = max_x; x >= min_x; x--)
```
```suggestion
		for(var/x in max_x to min_x)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+		for(var/x = min_x; x <= max_x; x++)
+			if( going & NORTH )
+				//Go along the Y axis from max to min, from up to down
+				for(var/y = max_y; y >= min_y; y--)
```
```suggestion
				for(var/y in max_y to min_y step -1)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+					lift_platform.travel(going)
+			else
+				//Go along the Y axis from min to max, from down to up
+				for(var/y = min_y; y <= max_y; y++)
```
```suggestion
				for(var/y in min_y to max_y)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+					lift_platform.travel(going)
+			else
+				//Go along the Y axis from min to max, from down to up
+				for(var/y = min_y; y <= max_y; y++)
```
```suggestion
				for(var/y in min_y to max_y)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+		for(var/x = max_x; x >= min_x; x--)
+			if( going & NORTH )
+				//Go along the Y axis from max to min, from up to down
+				for(var/y = max_y; y >= min_y; y--)
```
```suggestion
				for(var/y in max_y to min_y step -1)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+
+///Check destination turfs
+/datum/lift_master/proc/Check_lift_move(check_dir)
+	for(var/obj/structure/industrial_lift/lift_platform in lift_platforms)
+		var/turf/T = get_step_multiz(lift_platform, check_dir)
```
```suggestion
	for(var/lift_platform in lift_platforms)
		var/turf/T = get_step_multiz(lift_platform, check_dir)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	//	flags = CONDUCT_1
+	obj_flags = CAN_BE_HIT | BLOCK_Z_OUT_DOWN
+
+	var/list/lift_load = list() //things to move
```
```suggestion
	var/list/lift_load //things to move
```
lazy list

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	RegisterSignal(loc, COMSIG_ATOM_CREATED, .proc/AddItemOnLift)//For atoms created on platform
+
+/obj/structure/industrial_lift/proc/RemoveItemFromLift(datum/source, atom/movable/AM)
+	lift_load -= AM
```
```suggestion
	LAZYREMOVE(lift_load, AM)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	lift_load -= AM
+
+/obj/structure/industrial_lift/proc/AddItemOnLift(datum/source, atom/movable/AM)
+	lift_load |= AM
```
```suggestion
	LAZYOR(lift_load, AM)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+
+/obj/structure/industrial_lift/proc/lift_platform_expansion(datum/lift_master/LMaster)
+	. = list()
+	for(var/D in GLOB.cardinals)
+		var/turf/T = get_step(src, D)
+		. |= locate(/obj/structure/industrial_lift) in T
```
```suggestion
	for(var/direction in GLOB.cardinals)
		var/turf/neighbor = get_step(src, direction)
		. |= locate(/obj/structure/industrial_lift) in neighbor
```
Better variable names

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+		. |= locate(/obj/structure/industrial_lift) in T
+
+/obj/structure/industrial_lift/proc/travel(going)
+	var/list/things2move = lift_load.Copy()
```
```suggestion
	var/list/things2move = LAZYCOPY(lift_load)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	else
+		destination = going
+	forceMove(destination)
+	for(var/atom/movable/AM in things2move)
+		AM.forceMove(destination)
```
```suggestion
	for(var/am in things2move)
		var/atom/movable/thing = am
		thing.forceMove(destination)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	var/turf/can_move_up = LMaster.Check_lift_move(UP)
+	var/turf/can_move_up_down = LMaster.Check_lift_move(DOWN)
+
+	if (can_move_up || can_move_up_down)
+		var/result = show_radial_menu(user, src, tool_list, custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE, tooltips = TRUE)
+		if (!is_ghost && !in_range(src, user))
+			return  // nice try
+		switch(result)
+			if("Up")
+				if(can_move_up)
+					LMaster.MoveLift(UP, user)
+					show_fluff_message(TRUE, user)
+					use(user)
+				else
+					to_chat(user, "<span class='warning'>[src] doesn't seem to able move up!</span>")
+					use(user)
+			if("Down")
+				if(can_move_up_down)
+					LMaster.MoveLift(DOWN, user)
+					show_fluff_message(FALSE, user)
+					use(user)
+				else
+					to_chat(user, "<span class='warning'>[src] doesn't seem to able move down!</span>")
+					use(user)
+			if("Cancel")
+				return
+	else
+		to_chat(user, "<span class='warning'>[src] doesn't seem to able move anywhere!</span>")
+
+	add_fingerprint(user)
```
```suggestion
	if (!can_move_up && !can_move_up_down)
		to_chat(user, "<span class='warning'>[src] doesn't seem to able move anywhere!</span>")
		add_fingerprint(user)
		return

	var/result = show_radial_menu(user, src, tool_list, custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE, tooltips = TRUE)
	if (!is_ghost && !in_range(src, user))
		return  // nice try
	switch(result)
		if("Up")
			if(can_move_up)
				LMaster.MoveLift(UP, user)
				show_fluff_message(TRUE, user)
				use(user)
			else
				to_chat(user, "<span class='warning'>[src] doesn't seem to able move up!</span>")
				use(user)
		if("Down")
			if(can_move_up_down)
				LMaster.MoveLift(DOWN, user)
				show_fluff_message(FALSE, user)
				use(user)
			else
				to_chat(user, "<span class='warning'>[src] doesn't seem to able move down!</span>")
				use(user)
		if("Cancel")
			return
	
	add_fingerprint(user)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	QDEL_NULL(LMaster)
+	var/list/border_lift_platforms = lift_platform_expansion()
+	moveToNullspace()
+	for(var/obj/structure/industrial_lift/BorderLift in border_lift_platforms)
+		LMaster = new(BorderLift)
+	. = ..()
```
```suggestion
	for(var/border_lift in border_lift_platforms)
		LMaster = new(border_lift)
	return ..()
```
No need to type check
snake_case

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	if (!in_range(src, user))
+		return
+//NORTH, SOUTH, EAST, WEST, NORTHEAST, NORTHWEST, SOUTHEAST, SOUTHWEST
+	var/list/tool_list = list(
```
```suggestion
	var/static/list/tool_list = list(
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,266 @@
+	if (is_ghost && !in_range(src, user))
+		return
+
+	var/list/tool_list = list(
```
```suggestion
	var/static/list/tool_list = list(
```

Dennok committed:
lazy list

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
No type check

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
lazylist

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
return at the end of the proc

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
LAZYOR

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
No type check 2

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
length

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
Update code/game/objects/structures/industrial_lift.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
Update code/game/objects/structures/industrial_lift.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
another syntax of for

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
another syntax of for 2

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
another syntax of for 3

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
No type check 3

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
lazy list 2

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
lazy or

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
lazy remove

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
Better sanity

Better variable names and this way we don't fill the list with nulls in case of failure

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
no type 4

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
Update code/game/objects/structures/industrial_lift.dm

No need to type check
snake_case

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
static list

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
static list 2

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
Early returns

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
lazy copy

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
Update code/game/objects/structures/industrial_lift.dm

for

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
Update code/game/objects/structures/industrial_lift.dm

for

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
for 4

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok commented:
Great, i read about `batch` after this tonn of commits.

Dennok committed:
Update code/game/objects/structures/industrial_lift.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
lost tab

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,277 @@
+			var/list/result = borderline.lift_platform_expansion(src)
+			if(length(result))
+				for(var/p in result)
+					if(lift_platforms.Find(p))
```
```suggestion
					if(LAZYFIND(lift_platforms, p))
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,277 @@
+
+///Collect all bordered platforms
+/datum/lift_master/proc/Rebuild_lift_plaform(obj/structure/industrial_lift/base_lift_platform)
+	LAZYOR(lift_platforms, base_lift_platform)
```
Add and remove things from this list through procs, and register and unregister on their destruction, so if they get qdeleted early we don't keep a reference.
Examples:
```cpp
LAZYOR(lift_platforms, base_lift_platform)
```
becomes
```
add_lift_platforms(base_lift_platform)

/datum/lift_master/proc/add_lift_platforms(obj/structure/industrial_lift/new_lift_platform)
	if(new_lift_platform in lift_platforms)
		return
	LAZYADD(lift_platforms, new_lift_platform)
	RegisterSignal(new_lift_platform, COMSIG_PARENT_QDELETING, .proc/remove_lift_platforms)
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,277 @@
+	obj_flags = CAN_BE_HIT | BLOCK_Z_OUT_DOWN
+
+	var/list/lift_load //things to move
+	var/datum/lift_master/LMaster    //control from
```
```suggestion
	var/datum/lift_master/lift_master    //control from
```
Snake case

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,277 @@
+	//	flags = CONDUCT_1
+	obj_flags = CAN_BE_HIT | BLOCK_Z_OUT_DOWN
+
+	var/list/lift_load //things to move
```
```suggestion
	var/list/atom/movable/lift_load //things to move
```

Rohesie commented on code/game/objects/structures/industrial_lift.dm:
```diff
@@ -0,0 +1,277 @@
+	. = ..()
+	RegisterSignal(loc, COMSIG_ATOM_CREATED, .proc/AddItemOnLift)//For atoms created on platform
+
+/obj/structure/industrial_lift/proc/RemoveItemFromLift(datum/source, atom/movable/AM)
+	LAZYREMOVE(lift_load, AM)
+
+/obj/structure/industrial_lift/proc/AddItemOnLift(datum/source, atom/movable/AM)
+	LAZYOR(lift_load, AM)
+
```
```suggestion
/obj/structure/industrial_lift/proc/RemoveItemFromLift(datum/source, atom/movable/AM)
	if(!(AM in lift_load))
		return
	LAZYREMOVE(lift_load, AM)
	UnregisterSignal(AM, COMSIG_PARENT_QDELETING)

/obj/structure/industrial_lift/proc/AddItemOnLift(datum/source, atom/movable/AM)
	if(AM in lift_load)
		return
	LAZYADD(lift_load, AM)
	RegisterSignal(AM, COMSIG_PARENT_QDELETING, .proc/RemoveItemFromLift)

```

Dennok committed:
Up

Co-authored-by: Rohesie <rohesie@gmail.com>

Dennok committed:
Some Up

Dennok committed:
ups

Rohesie commented:
Failing travis.

Dennok commented:
I don't understand why it failing.

Rohesie commented:
```
code\game\objects\structures\industrial_lift.dm:113:error: smooth: undefined var
code\game\objects\structures\industrial_lift.dm:113:error: SMOOTH_MORE: undefined var
```
There has been a smoothing refactor, and these variables are outdated now.

Dennok committed:
Merge pull request #18 from tgstation/master

Up upstream

Dennok commented:
Yes, thanks.

Dennok committed:
lift smooth flag

Dennok committed:
up smoothing

Dennok commented:
Finaly, it pass travis.

Dennok commented:
Great, now i try make a console for it.
